include/lbann/models/model.hpp

```diff
@@ -125,11 +117,20 @@ class model {
     return m_metrics;
   }
 
-  /** Set the model's layers. */
-  void set_layers(std::vector<Layer *>& layers);
-
-  /** Return the model's layers. */
-  virtual const std::vector<Layer *>& get_layers() const { return m_layers; }
+  /** Size of model's list of layers. */
+  El::Int get_num_layers() const noexcept;
+  /** @param pos Position in model's list of layers. */
+  Layer& get_layer(El::Int pos);
+  /** @param pos Position in model's list of layers. */
+  const Layer& get_layer(El::Int pos) const;
+  /** @brief Return list of layers in model.
+   *  @details The list is in execution order for forward propagation.
+   */
+  std::vector<Layer*> get_layers();
```

Comment:
I have some questions about the paradigm here.  If we are using unique pointers do we want to be returning raw pointers for this function.

Reply:
Ideally what you want is `std::vector<Layer&>`, but that's not an easy (or sane) thing to create. The semantics is `std::vector<SomeObserver<Layer>>` --> `std::vector<observing_ptr<layer>>` --> `std::vector<layer*>`. I think this is the best we can do. Even better, though, would be to support iterator access only, so this function can go away completely.